### PR TITLE
(CTH-390) Add PXP agent init scripts

### PR DESCRIPTION
**Pull request by geoffnichols**
This commit adds init scripts and service config
files for initialization of the PXP agent service.

These are derived as wholly as possible from the
facilities currently used for Puppet in the AIO
puppet-agent.

### CODE REVIEW

**Comment by geoffnichols** [comment]
Replaces https://github.com/puppetlabs/pxp-agent/pull/96.

**Comment by geoffnichols** [comment]
Test build in progress: http://jenkins-release.delivery.puppetlabs.net/job/vanagon_generic_job/143/.

**Comment by haus** [comment]
Style nitpick. The redhat script has a mix of 2 and 4 space tabs, while deb and sles have 4 space tabs.

File: ext/osx/pxp-agent.plist [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+<dict>
+        <key>EnvironmentVariables</key>
+        <dict>
+                <key>PATH</key>
+                <string>/opt/puppetlabs/puppet/bin:/usr/bin:/bin:/usr/sbin:/sbin</string>
+                <key>LANG</key>
+                <string>en_US.UTF-8</string>
+        </dict>
+        <key>Label</key>
+        <string>pxp-agent</string>
```
Comment (geoffnichols) [inline ext/osx/pxp-agent.plist]: There are open questions in https://github.com/puppetlabs/pxp-agent/pull/96 on whether this should be fully qualified.

**Comment by geoffnichols** [comment]
Updated commit to improve consistency of indentation in the redhat script.

File: ext/debian/pxp-agent.init [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+pidfile="/var/run/puppetlabs/${prog}.pid"
+
+[ -x $exec ] || exit 5
+
+[ -r /etc/default/pxp-agent ] && . /etc/default/pxp-agent
+
+. /lib/lsb/init-functions
+
+reload_pxp_agent() {
+    start-stop-daemon --stop --quiet --signal HUP --pidfile $pidfile --name $prog
```
Comment (haus) [inline ext/debian/pxp-agent.init]: do we know if pxp responds to HUP?

Comment (haus) [inline ext/debian/pxp-agent.init]: `signal(SIGHUP, SIG_IGN);`

So HUP is currently ignored and marked as a TODO in the daemonize code. I think it is misleading to offer a reload method that we know won't work. Maybe leave the function, but drop it out of the list at the end?

Comment (kylog) [inline ext/debian/pxp-agent.init]: Yep that makes sense. And we can add a ticket adding a SIGHUP handler for realz.

File: ext/osx/pxp-agent.plist (hunk 1 above)
Comment (kylog) [inline ext/osx/pxp-agent.plist]: Specifically at https://github.com/puppetlabs/pxp-agent/pull/96/files#r40988986. We should resolve that one way or the other. Ping @stahnma or @heathseals.

**Comment by geoffnichols** [comment]
Update commit to remove `reload` and `force-reload` from the "Usage:" list.

File: ext/systemd/pxp-agent.service [hunk 3]
```diff
@@ -0,0 +1,12 @@
+[Unit]
+Description=PCP Execution Protocol (PXP) Agent
+After=syslog.target network.target
+
+[Service]
+EnvironmentFile=-/etc/sysconfig/pxp-agent
+EnvironmentFile=-/etc/default/pxp-agent
+ExecReload=/bin/kill -HUP $MAINPID
```
Comment (haus) [inline ext/systemd/pxp-agent.service]: Not sure if we should leave this or pull it out.

File: ext/suse/pxp-agent.init [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+# 5 - program not installed
+# 6 - program not configured
+#
+# Note that starting an already running service, stopping
+# or restarting a not-running service as well as the restart
+# with force-reload (in case signalling is not supported) are
+# considered a success.
+
+case "$1" in
+    start)
```
Comment (haus) [inline ext/suse/pxp-agent.init]: the rpm script makes the pid dir if it doesn't exist already (for tmpfs systems), we probably want to do the same here, unless startproc makes the dir for us.

File: ext/debian/pxp-agent.init [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+start_pxp_agent() {
+    start-stop-daemon --start --quiet --pidfile $pidfile --startas $exec -- --daemonize $PXP_AGENT_OPTIONS
```
Comment (haus) [inline ext/debian/pxp-agent.init]: the rpm script makes the pid dir if it doesn't exist already (for tmpfs systems), we probably want to do the same here, unless start-stop-daemon makes the dir for us.

**Comment by haus** [comment]
@geoffnichols also because of the way the aix service works in vanagon, that particular file has to live in puppet-agent. i'm so sorry.

File: ext/debian/pxp-agent.init [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+[ -r /etc/default/pxp-agent ] && . /etc/default/pxp-agent
+
+. /lib/lsb/init-functions
+
+reload_pxp_agent() {
+    start-stop-daemon --stop --quiet --signal HUP --pidfile $pidfile --name $prog
+}
+
+start_pxp_agent() {
+    mkdir -p $piddir
```
Comment (geoffnichols) [inline ext/debian/pxp-agent.init]: Updated to ensure $piddir exists before we start the daemon.

File: ext/suse/pxp-agent.init [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        ## of the pidfile, and do nothing if true. Otherwise
+        ## force a process start
+        if [ -f "${pidfile}" ]; then
+            PID=$(cat "$pidfile")
+            if [ "$PID" -eq $(pgrep -f "$exec") ] ; then
+                rc_status -v
+                rc_exit
+            fi
+        fi
+        mkdir -p $piddir
```
Comment (geoffnichols) [inline ext/suse/pxp-agent.init]: Updated to ensure $piddir exists before we start the daemon.

File: ext/systemd/pxp-agent.service (hunk 3 above)
Comment (geoffnichols) [inline ext/systemd/pxp-agent.service]: I'm not sure either.

**Comment by geoffnichols** [comment]
@haus Thanks, I moved the aix service into the corresponding puppet-agent [PR]( https://github.com/puppetlabs/puppet-agent/pull/321/files#diff-c58fa82954e16bcd52c85eaca487b0e8R1).

**Comment by haus** [comment]
LGTM :+1:. Have not tested.

**Comment by geoffnichols** [comment]
Another test build is in progress: http://jenkins-release.delivery.puppetlabs.net/job/vanagon_generic_job/146/.

File: ext/osx/pxp-agent.plist (hunk 1 above)
Comment (shrug) [inline ext/osx/pxp-agent.plist]: If we're still planning to move the other services to the fully qualified form (we moved them before but it caused some issues so we reverted, needed more work), then we should go ahead and implement this one, IMO.

Comment (kylog) [inline ext/osx/pxp-agent.plist]: FWIW my $.02 is: it sounds like we don't have the work scoped for this change, and also consistency is nice to have. So for now, I'd leave this one unqualified (i.e. just `pxp-agent`), and create a ticket (or tickets) for updating all the agent services to fully-qualified names and making whatever related changes are needed in, say, the enterprise modules managing these services (and other modules?).

Thoughts?

Comment (geoffnichols) [inline ext/osx/pxp-agent.plist]: @jpartlow @ericwilliamson Please see linked discussion above. Any thoughts on implications of "fully qualifying" the  OS X launchd service for pxp-agent?

Comment (shrug) [inline ext/osx/pxp-agent.plist]: Fair enough, that works for me. Consistency is good

File: ext/systemd/pxp-agent.service (hunk 3 above)
Comment (shrug) [inline ext/systemd/pxp-agent.service]: From https://fedoraproject.org/wiki/Packaging:Systemd
"ExecReload= should be specified for all services supporting reload. It is highly recommended to add code here that synchronously reloads the configuration file here (i.e. /bin/kill -HUP $MAINPID is usually a poor choice, due to its asynchronous nature). Omit this option if your service does not support reloading."

**Comment by geoffnichols** [comment]
Updated commit to remove `ExecReload=/bin/kill -HUP $MAINPID` from the systemd service file.

**Comment by shrug** [comment]
:+1:

**Pull request merged**